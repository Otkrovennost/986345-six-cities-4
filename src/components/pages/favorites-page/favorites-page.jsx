import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import Header from "../../header/header.jsx";
import FavoritesEmpty from "../../favorites-empty/favorites-empty.jsx";
import Card from "../../card/card.jsx";
import {CardClass} from "../../../const.js";
import {getFavoriteOffers, getCitiesOffersList, getFavoriteOffersStatus} from "../../../reducer/data/selectors.js";
import {Operation as DataOperation} from "../../../reducer/data/data.js";

class FavoritesPage extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {loadFavoriteOffers} = this.props;
    loadFavoriteOffers();
  }

  render() {
    const {favoriteOffers, citiesOffersList, isFavoriteOffersLoading} = this.props;

    if (isFavoriteOffersLoading) {
      return false;
    } else if (!favoriteOffers.length) {
      return <FavoritesEmpty />;
    }

    return (
      <div className="page page--favorites-empty">
        <Header />
        <main className="page__main page__main--favorites">
          <div className="page__favorites-container container">
            <section className="favorites">
              <h1 className="favorites__title">Saved listing</h1>
              <ul className="favorites__list">
                {citiesOffersList.map((city) => {
                  return (
                    favoriteOffers.filter((offer) => offer.city === city).length !== 0 ?
                      <li className="favorites__locations-items" key={city}>
                        <div className="favorites__locations locations locations--current">
                          <div className="locations__item">
                            <a className="locations__item-link" href="#">
                              <span>{city}</span>
                            </a>
                          </div>
                        </div>
                        <div className="favorites__places">
                          {favoriteOffers.filter((offer) => offer.city === city).map((card) => {
                            return (
                              <Card
                                key={card.id}
                                offer={card}
                                cardClass={CardClass.FAVORITES}
                              />);
                          })}
                        </div>
                      </li> : ``
                  );
                }
                )}
              </ul>
            </section>
          </div>
        </main>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  favoriteOffers: getFavoriteOffers(state),
  citiesOffersList: getCitiesOffersList(state),
  isFavoriteOffersLoading: getFavoriteOffersStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadFavoriteOffers() {
    dispatch(DataOperation.loadFavoriteOffers());
  }
});

FavoritesPage.propTypes = {
  loadFavoriteOffers: PropTypes.func,
  favoriteOffers: PropTypes.array,
  citiesOffersList: PropTypes.array,
  isFavoriteOffersLoading: PropTypes.bool
};

export {FavoritesPage};
export default connect(mapStateToProps, mapDispatchToProps)(FavoritesPage);
