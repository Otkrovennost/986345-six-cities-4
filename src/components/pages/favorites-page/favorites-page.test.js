import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {BrowserRouter} from "react-router-dom";
import NameSpace from "../../../reducer/name-space.js";
import {FavoritesPage} from "./favorites-page.jsx";

const mockStore = configureStore([]);
const CITIES_LIST = [`Amsterdam`, `Paris`, `Cologne`, `Brussels`, `Hamburg`];
const FAVORITE_OFFERS = [{
  id: 1,
  city: `Amsterdam`,
  title: `Beautiful & luxurious apartment at great location`,
  coords: [52.3909553943508, 4.85309666406198],
  description: ` A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
  price: 120,
  rating: 4.8,
  type: `Apartment`,
  photo: `img/apartment-01.jpg`,
  isPremium: true,
  bookmark: false,
  quantityBedrooms: 3,
  maxAdults: 4,
  options: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    id: 1,
    isSuper: true,
    name: `Angelina`
  }
}];

it(`Should Favorites component render correctly`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      citiesOffersList: CITIES_LIST,
      favoriteOffers: FAVORITE_OFFERS,
      isFavoriteOffersLoading: true
    },
    [NameSpace.USER]: {
      isSignIn: true
    }
  });

  const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <FavoritesPage
            loadFavoriteOffers={() => {}}
          />
        </BrowserRouter>
      </Provider>
  )
        .toJSON();

  expect(tree).toMatchSnapshot();
});
