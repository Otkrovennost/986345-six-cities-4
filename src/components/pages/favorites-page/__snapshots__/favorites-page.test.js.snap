// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Favorites component render correctly 1`] = `null`;
